crates/hir_ty/src/infer/expr.rs

```diff
@@ -358,7 +358,7 @@ impl<'a> InferenceContext<'a> {
 
                 Ty::simple(TypeCtor::Never)
             }
-            Expr::Return { expr } => {
+            Expr::Return { expr } | Expr::Yield { expr } => {
                 if let Some(expr) = expr {
                     self.infer_expr_coerce(*expr, &Expectation::has_type(self.return_ty.clone()));
```

Comment:
This doesn't apply to yield I think. A generator may have a return type and a yield type that aren't the same. To properly implement inference for `yield` we need to track the yield type in the `InferenceContext` I believe but that can be done in a separate PR I'd say.
So for now we probably want something like this:
```rust
Expr::Yield { expr } => {
	  // FIXME: track yield type for coercion
	  if let Some(expr) = expr {
	      self.infer_expr(expr, &Expectation::none());
	  }
	  Ty::simple(TypeCtor::Never)
}
```